Support for MacOS Darwin

noelmcloughlin opened the pull request:
This PR introduces support for Darwin.  tested on MacOS High Sierra.

```
[WARNING ] The function "module.run" is using its deprecated version and will expire in version "Sodium".
[ERROR   ] Command '/usr/libexec/PlistBuddy -c "print :CFBundleIdentifier" ''"'"'/tmp/dmg-ThzeS9/*.pkg'"'"'/Contents/Info.plist'' failed with return code: 1
[ERROR   ] output: Print: Entry, ":CFBundleIdentifier", Does Not Exist
File Doesn't Exist, Will Create: '/tmp/dmg-ThzeS9/*.pkg'/Contents/Info.plist
[WARNING ] The function "module.run" is using its deprecated version and will expire in version "Sodium".
[ERROR   ] {u'ret': False}
local:
----------
          ID: java-install-dir
    Function: file.directory
        Name: /Library/Java/JavaVirtualMachines
      Result: True
     Comment: The directory /Library/Java/JavaVirtualMachines is in the correct state
     Started: 19:35:25.068538
    Duration: 0.835 ms
     Changes:   
----------
          ID: sun-java-remove-prev-archive
    Function: file.absent
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: File /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg is not present
     Started: 19:35:25.069708
    Duration: 0.244 ms
     Changes:   
----------
          ID: download-jdk-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg' 'http://download.oracle.com/otn-pub/java/jdk/8u172-b11/a58eab1ec242421181065cdc37240b08/jdk-8u172-macosx-x64.dmg'
      Result: True
     Comment: Command "curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg' 'http://download.oracle.com/otn-pub/java/jdk/8u172-b11/a58eab1ec242421181065cdc37240b08/jdk-8u172-macosx-x64.dmg'" run
     Started: 19:35:25.070736
    Duration: 636534.922 ms
     Changes:   
              ----------
              pid:
                  83058
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: check-jdk-archive
    Function: module.run
        Name: file.check_hash
      Result: True
     Comment: Module function file.check_hash executed
     Started: 19:46:01.607533
    Duration: 656.862 ms
     Changes:   
              ----------
              ret:
                  True
----------
          ID: unpack-jdk-archive
    Function: archive.extracted
        Name: /Library/Java/JavaVirtualMachines
      Result: True
     Comment: unless condition is true
     Started: 19:46:02.265975
    Duration: 3100.605 ms
     Changes:   
----------
          ID: unpack-jdk-macpackage
    Function: macpackage.installed
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg installed
     Started: 19:46:05.367073
    Duration: 8950.996 ms
     Changes:   
              ----------
              installed:
----------
          ID: update-javahome-symlink
    Function: file.symlink
        Name: /usr/local/lib/java
      Result: True
     Comment: Created new symlink /usr/local/lib/java -> /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home
     Started: 19:46:14.319464
    Duration: 28.007 ms
     Changes:   
              ----------
              new:
                  /usr/local/lib/java
----------
          ID: remove-jdk-archive
    Function: file.absent
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: Removed file /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
     Started: 19:46:14.347973
    Duration: 15.685 ms
     Changes:   
              ----------
              removed:
                  /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
----------
          ID: sun-java-jce-unzip
    Function: pkg.installed
        Name: unzip
      Result: True
     Comment: All specified packages are already installed
     Started: 19:46:14.372324
    Duration: 641.787 ms
     Changes:   
----------
          ID: sun-java-remove-old-jce-archive
    Function: file.absent
        Name: /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: True
     Comment: File /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip is not present
     Started: 19:46:15.014884
    Duration: 0.457 ms
     Changes:   
----------
          ID: download-jce-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip' 'http://download.oracle.com/otn-pub/java/jdk/8/jce_policy-8.zip'
      Result: True
     Comment: Command "curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip' 'http://download.oracle.com/otn-pub/java/jdk/8/jce_policy-8.zip'" run
     Started: 19:46:15.016123
    Duration: 3218.044 ms
     Changes:   
              ----------
              pid:
                  84275
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: check-jce-archive
    Function: module.run
        Name: file.check_hash
      Result: False
     Comment: Module function file.check_hash executed
     Started: 19:46:18.235339
    Duration: 1.588 ms
     Changes:   
              ----------
              ret:
                  False
----------
          ID: backup-non-jce-jar
    Function: cmd.run
        Name: mv US_export_policy.jar US_export_policy.jar.nonjce; mv local_policy.jar local_policy.jar.nonjce;
      Result: False
     Comment: One or more requisite failed: sun-java.jce.check-jce-archive
     Changes:   
----------
          ID: unpack-jce-archive
    Function: cmd.run
        Name: unzip -j -o /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: False
     Comment: One or more requisite failed: sun-java.jce.backup-non-jce-jar, sun-java.jce.check-jce-archive
     Changes:   
----------
          ID: remove-jce-archive
    Function: file.absent
        Name: /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: False
     Comment: One or more requisite failed: sun-java.jce.unpack-jce-archive
     Changes:   
----------
          ID: jdk-config
    Function: file.managed
        Name: /etc/profile.d/java.sh
      Result: True
     Comment: unless condition is true
     Started: 19:46:18.238533
    Duration: 16.226 ms
     Changes:   
----------
          ID: javahome-link
    Function: file.symlink
        Name: /usr/local/lib/java
      Result: True
     Comment: Symlink /usr/local/lib/java is present and owned by root:wheel
     Started: 19:46:18.255272
    Duration: 28.964 ms
     Changes:   
----------
          ID: java-link
    Function: file.symlink
        Name: /usr/local/bin/java
      Result: True
     Comment: Created new symlink /usr/local/bin/java -> /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home/bin/java
     Started: 19:46:18.284843
    Duration: 71.504 ms
     Changes:   
              ----------
              new:
                  /usr/local/bin/java
----------
          ID: javac-link
    Function: file.symlink
        Name: /usr/local/bin/javac
      Result: True
     Comment: Created new symlink /usr/local/bin/javac -> /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home/bin/javac
     Started: 19:46:18.362174
    Duration: 2862.623 ms
     Changes:   
              ----------
              new:
                  /usr/local/bin/javac

Summary for local
-------------
Succeeded: 15 (changed=9)
Failed:     4
```

vutny commented on README.rst:
```diff
@@ -8,6 +8,7 @@ Formula to set up and configure Java JREs and JDKs from a tarball archive source
 
     See the full `Salt Formulas installation and usage instructions
     <http://docs.saltstack.com/en/latest/topics/development/conventions/formulas.html>`_.
+    Verified on Linux and MacOS.
```
I think you might wanted to add it at the end of the file where tested OS are listed.

vutny commented on sun-java/init.sls:
```diff
@@ -55,10 +55,28 @@ unpack-jdk-archive:
+    - app: False
+    - force: True
+    - allow_untrusted: True
+    - onlyif: test "`uname`" = "Darwin"
```
Although this is technically OK and should work fine, but I'd rather use Jinja conditional statement to render different states for MacOS and Linux. Since `macpackage` state module could not be loaded on Linux platform, this may cause some silly bugs. Just better to avoid that.

noelmcloughlin commented:
I've implemented the suggestions. Thanks @vutny 
----
```
[WARNING ] The function "module.run" is using its deprecated version and will expire in version "Sodium".
[ERROR   ] Command '/usr/libexec/PlistBuddy -c "print :CFBundleIdentifier" ''"'"'/tmp/dmg-D8Zn3G/*.pkg'"'"'/Contents/Info.plist'' failed with return code: 1
[ERROR   ] output: Print: Entry, ":CFBundleIdentifier", Does Not Exist
File Doesn't Exist, Will Create: '/tmp/dmg-D8Zn3G/*.pkg'/Contents/Info.plist
[WARNING ] The function "module.run" is using its deprecated version and will expire in version "Sodium".
[ERROR   ] {u'ret': False}
local:
----------
          ID: java-install-dir
    Function: file.directory
        Name: /Library/Java/JavaVirtualMachines
      Result: True
     Comment: The directory /Library/Java/JavaVirtualMachines is in the correct state
     Started: 14:15:18.408057
    Duration: 0.803 ms
     Changes:   
----------
          ID: sun-java-remove-prev-archive
    Function: file.absent
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: File /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg is not present
     Started: 14:15:18.409497
    Duration: 0.273 ms
     Changes:   
----------
          ID: download-jdk-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg' 'http://download.oracle.com/otn-pub/java/jdk/8u172-b11/a58eab1ec242421181065cdc37240b08/jdk-8u172-macosx-x64.dmg'
      Result: True
     Comment: Command "curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg' 'http://download.oracle.com/otn-pub/java/jdk/8u172-b11/a58eab1ec242421181065cdc37240b08/jdk-8u172-macosx-x64.dmg'" run
     Started: 14:15:18.411224
    Duration: 693166.186 ms
     Changes:   
              ----------
              pid:
                  128
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: check-jdk-archive
    Function: module.run
        Name: file.check_hash
      Result: True
     Comment: Module function file.check_hash executed
     Started: 14:26:51.580743
    Duration: 579.345 ms
     Changes:   
              ----------
              ret:
                  True
----------
          ID: unpack-jdk-archive
    Function: macpackage.installed
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg installed
     Started: 14:26:55.161722
    Duration: 9346.295 ms
     Changes:   
              ----------
              installed:
----------
          ID: update-javahome-symlink
    Function: file.symlink
        Name: /usr/local/lib/java
      Result: True
     Comment: Symlink /usr/local/lib/java is present and owned by root:wheel
     Started: 14:27:04.508369
    Duration: 17.844 ms
     Changes:   
----------
          ID: remove-jdk-archive
    Function: file.absent
        Name: /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
      Result: True
     Comment: Removed file /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
     Started: 14:27:04.526786
    Duration: 12.74 ms
     Changes:   
              ----------
              removed:
                  /Library/Java/JavaVirtualMachines/jdk-8u172-macosx-x64.dmg
----------
          ID: sun-java-jce-unzip
    Function: pkg.installed
        Name: unzip
      Result: True
     Comment: All specified packages are already installed
     Started: 14:27:04.551506
    Duration: 901.663 ms
     Changes:   
----------
          ID: sun-java-remove-old-jce-archive
    Function: file.absent
        Name: /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: True
     Comment: File /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip is not present
     Started: 14:27:05.453870
    Duration: 0.44 ms
     Changes:   
----------
          ID: download-jce-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip' 'http://download.oracle.com/otn-pub/java/jdk/8/jce_policy-8.zip'
      Result: True
     Comment: Command "curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip' 'http://download.oracle.com/otn-pub/java/jdk/8/jce_policy-8.zip'" run
     Started: 14:27:05.454696
    Duration: 16476.183 ms
     Changes:   
              ----------
              pid:
                  1599
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: check-jce-archive
    Function: module.run
        Name: file.check_hash
      Result: False
     Comment: Module function file.check_hash executed
     Started: 14:27:21.931946
    Duration: 1.52 ms
     Changes:   
              ----------
              ret:
                  False
----------
          ID: backup-non-jce-jar
    Function: cmd.run
        Name: mv US_export_policy.jar US_export_policy.jar.nonjce; mv local_policy.jar local_policy.jar.nonjce;
      Result: False
     Comment: One or more requisite failed: sun-java.jce.check-jce-archive
     Changes:   
----------
          ID: unpack-jce-archive
    Function: cmd.run
        Name: unzip -j -o /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: False
     Comment: One or more requisite failed: sun-java.jce.backup-non-jce-jar, sun-java.jce.check-jce-archive
     Changes:   
----------
          ID: remove-jce-archive
    Function: file.absent
        Name: /usr/local/lib/java/jre/lib/security/policy/unlimited/UnlimitedJCEPolicy.zip
      Result: False
     Comment: One or more requisite failed: sun-java.jce.unpack-jce-archive
     Changes:   
----------
          ID: jdk-config
    Function: file.managed
        Name: /etc/profile.d/java.sh
      Result: True
     Comment: unless condition is true
     Started: 14:27:21.934977
    Duration: 526.725 ms
     Changes:   
----------
          ID: javahome-link
    Function: file.symlink
        Name: /usr/local/lib/java
      Result: True
     Comment: Symlink /usr/local/lib/java is present and owned by root:wheel
     Started: 14:27:22.462079
    Duration: 17.08 ms
     Changes:   
----------
          ID: java-link
    Function: file.symlink
        Name: /usr/local/bin/java
      Result: True
     Comment: Symlink /usr/local/bin/java is present and owned by root:wheel
     Started: 14:27:22.479757
    Duration: 29.822 ms
     Changes:   
----------
          ID: javac-link
    Function: file.symlink
        Name: /usr/local/bin/javac
      Result: True
     Comment: Symlink /usr/local/bin/javac is present and owned by root:wheel
     Started: 14:27:22.510237
    Duration: 29.599 ms
     Changes:   

Summary for local
-------------
Succeeded: 14 (changed=6)
Failed:     4
-------------
Total states run:     18
Total run time:  721.107 s
```

noelmcloughlin commented:
Thanks for reviewing @vutny
